### ISSUE TITLE
ConnectionClosed error in .transfer_data

### ISSUE THREAD

**Issue by lgrahl** [aaugustin/websockets#606]
The library logs an error in case a ConnectionClosed is being raised within `.transfer_data`.

Version: 7.0

```python
Error in data transfer
Traceback (most recent call last):
  File "/websockets/protocol.py", line 674, in transfer_data
    message = yield from self.read_message()
  File "/websockets/protocol.py", line 742, in read_message
    frame = yield from self.read_data_frame(max_size=self.max_size)
  File "/websockets/protocol.py", line 825, in read_data_frame
    yield from self.write_close_frame(frame.data)
  File "/websockets/protocol.py", line 958, in write_close_frame
    True, OP_CLOSE, data, _expected_state=State.CLOSING
  File "/websockets/protocol.py", line 919, in write_frame
    yield from self.ensure_open()
  File "/websockets/protocol.py", line 646, in ensure_open
    ) from self.transfer_data_exc
websockets.exceptions.ConnectionClosed: WebSocket connection is closed: code = 1001 (going away), no reason
```

Should this be silently discarded?

**Comment by aaugustin** [comment]
Unsure about this one...

Do you have a bit more context? Can you show the code that causes this?

**Comment by lgrahl** [comment]
This happens daily on our production [SaltyRTC Server](https://github.com/saltyrtc/saltyrtc-server-python/tree/v4.2.0) instance with moderate load. Can't really tell why it occurs, so I don't know how to reproduce it. From our log context surrounding the error, everything was normal and we eventually got a `ConnectionClosed`.

**Comment by aaugustin** [comment]
Based on the stack trace, this happens when:

- In `transfer_data, `websockets` receives a close frame with code 1001, which is what browsers do when navigating away from a page where a WebSocket connection is open, and responds by echoing the close frame.
- In `write_frame`, `websockets` sends that frame and waits for the connection to drain, but that fails with a `ConnectionError` — presumably because the remote endpoint closed the connection without waiting for the response.
- `write_frame` raises the appropriate `ConnectionClosed` exception.

Actually, when `transfer_data` sends back the close frame, it doesn't really care that it gets send. I think it should ignore `ConnectionClosed` exceptions.

**Comment by aaugustin** [comment]
The fix should be:

```diff
diff --git a/src/websockets/protocol.py b/src/websockets/protocol.py
index d6462cc..d888a97 100644
--- a/src/websockets/protocol.py
+++ b/src/websockets/protocol.py
@@ -864,10 +864,15 @@ class WebSocketCommonProtocol(asyncio.StreamReaderProtocol):
                 # 7.1.5.  The WebSocket Connection Close Code
                 # 7.1.6.  The WebSocket Connection Close Reason
                 self.close_code, self.close_reason = parse_close(frame.data)
-                # Echo the original data instead of re-serializing it with
-                # serialize_close() because that fails when the close frame is
-                # empty and parse_close() synthetizes a 1005 close code.
-                await self.write_close_frame(frame.data)
+                try:
+                    # Echo the original data instead of re-serializing it with
+                    # serialize_close() because that fails when the close frame
+                    # is empty and parse_close() synthetizes a 1005 close code.
+                    await self.write_close_frame(frame.data)
+                except ConnectionClosed:
+                    # It doesn't really matter if the connection was closed
+                    # before we could send back a close frame.
+                    pass
                 return None

             elif frame.opcode == OP_PING:
```

It needs a test.

**Comment by lgrahl** [comment]
:+1: I'll double-check it is resolved once 8.0 has been released.
